### PR TITLE
Fast

### PR DESCRIPTION
**Pull request by forthy42**
# All my tuning together

Ok, here's everything I so far wanted to do in one big lump of patchset.

1. errno-style error numbering
2. two-stage glyph and kerning lookup instead of linear search
3. avoid reloading fonts for every glyph

I've done strace on the demos/font example, and rendering takes 0.304s with the original "one open call per single glyph", and 0.0798s with "one open call per glyphs batch (one size)" as is now the default. The interface is backwards compatible, i.e. there is now a per-font mode when to open and close the file, and there are two new calls

texture_font_default_mode(font, [ MODE_AUTO_CLOSE, MODE_GLYPHS_CLOSE, MODE_FREE_CLOSE, MODE_MANUAL_CLOSE, MODE_ALWAYS_OPEN ])

and

texture_font_close(font, mode_for_font, mode_for_library)

The normal default for a font is MODE_AUTO_CLOSE, which will be raised to MODE_GLYPHS_CLOSE in texture_font_load_glyphs, and so mostly emulate the current behavior of open and freeing the font face for every glyph rendered.  The default mode for the FT_Library is MODE_ALWAYS_OPEN, I don't see a point in closing and opening that structure.  It's now a per-thread structure reused for every font created in the same thread, so if you want to render several texture atlases in parallel, make sure you create the font faces in the same thread you use it in to render glyphs.

What I still want to add is a mode for glyph texture coordinates, where the coordinates are stored as pixels, and the scaling is done in the shader program (optional feature).  That allows to not worry about multiple fonts being in one atlas, and needing to resize all the glyphs when the atlas is full.  Just update your shader parameter.

And a "needs synced with the texture" flag in the atlas, to avoid needless syncs.  That's both quick to do, and I'll probably do it this evening.

The CMakeLists.txt now contains a configurable debug/release option, and also builds shared libraries.  I have inserted some dummy version numbers (shared libraries need those), as I have no clue what version name I should start with ;-).

I've not merged in all changes into the harfbuzz stuff, and I really think the harfbuzz support should be an additional call, not a separate library with just one file being different, and even that file is mostly the same stuff, and causes maintenance problems when anything changes.

### CODE REVIEW

**Comment by adrianbroher** [comment]
>  in one big lump of patchset

Not interested at all.

**Comment by forthy42** [comment]
I can easily separate the individual improvements into separate patchsets, but my errno pull request is still pending, and I don't want to create non-mergeable patchsets... If you want to merge them one-by-one, merge the errno patchset first.

**Comment by AqilCont** [comment]
This looks cool, will it never be merged?

**Comment by rougier** [comment]
Sorry, I just forgot about this PR... @forthy42 Still interested in having it merged?

**Comment by forthy42** [comment]
Yes, still interested.

**Comment by rougier** [comment]
Nice. I guess there are some changes needed to the merging of PR since yours. Could you have a look to check it this would require only cosmectic change or a complete overhaul of your PR?

**Comment by forthy42** [comment]
I'll have a look at the merge conflicts.

**Comment by forthy42** [comment]
So far, looks like mostly cosmetic changes.

**Comment by michaeltyson** [comment]
I just made my own pull request that avoids loading the font for every glyph before I saw this PR. I'd love to see this merged in, @rougier

**Comment by AqilCont** [comment]
Yeah, me too! Faster font loading is always welcomed.

**Comment by rougier** [comment]
I think a rebase is needed.

**Comment by michaeltyson** [comment]
> I think a rebase is needed.

Up for that, @forthy42 ? I could give it a crack too, but then I think it would complicate things because I’d have to create a separate PR. Hopefully third time’s the charm eh? 😄

**Comment by forthy42** [comment]
I think I have time for this today or tomorrow.

**Comment by forthy42** [comment]
Ok, one major source of conflict was double work: I improved the error handling, and you improved the error handling. The merge now has no conflicts, but fails the checks.

**Comment by michaeltyson** [comment]
Great, thanks for doing that! Have you got time to sort out the remaining stuff with the checks, or shall I dive in? (I think that'll mean forking your fork, and making my own composite PR once I'm done, though)

**Comment by forthy42** [comment]
Checks on Windows still fail. Problems: Doesn't find snprintf (should be in stdio.h), doesn't find <endian.h> (probably Linux-only).

**Comment by forthy42** [comment]
No conflicts, checks pass.

**Comment by forthy42** [comment]
> Great, thanks for doing that! Have you got time to sort out the remaining stuff with the checks, or shall I dive in? (I think that'll mean forking your fork, and making my own composite PR once I'm done, though)

My PR includes emoji support; you can check if your PR does something better.

The load&close font for every glyph code is not well tested anymore, I suppose there are issues with that.  If anyone wants to keep that mode, test it; otherwise the next PR should remove this mode.

**Comment by michaeltyson** [comment]
Cool, nicely done! I've just done a PR against your repo with some additional fixes. Trying out the emoji support next, if I can figure it out 😄

File: texture-font.c [hunk 1]
```diff
@@ -736,14 +787,22 @@ texture_font_load_glyph( texture_font_t * self,
...
-    FT_Load_Glyph( self->face, glyph_index, FT_LOAD_RENDER | FT_LOAD_NO_HINTING);
+    FT_Load_Glyph( self->face, glyph_index,
+		   ((self->atlas->depth == 4) ? FT_LOAD_COLOR : 0) |
+		   FT_LOAD_RENDER | FT_LOAD_NO_HINTING);
     slot = self->face->glyph;
-    glyph->advance_x = slot->advance.x / HRESf;
-    glyph->advance_y = slot->advance.y / HRESf;
+    if( self->atlas->depth == 4 ) {
+	// color fonts use actual pixels, not subpixels
+	glyph->advance_x = slot->advance.x * self->scale;
```
Comment (michaeltyson) [inline texture-font.c]: This part seems to be messing up my normal text rendering - I'm using an atlas with depth 4 so I can store ordinary text and emojis in the same atlas, and thus render in the same pass. If I use the calculation in the "else" block, then rendering is fine.

File: texture-font.c [hunk 2]
```diff
@@ -511,9 +542,16 @@ texture_font_delete( texture_font_t *self )
 {
     size_t i;
     texture_glyph_t *glyph;
+    FT_Error error=0;
 
     assert( self );
 
+    error = FT_Done_Size( self->ft_size );
```
Comment (michaeltyson) [inline texture-font.c:563]: Crash here if texture_font_close is called before this point, which can happen if a load fails (happening to me right now as the emoji support isn't quite working yet).

Perhaps wrap it in a check:

`
    if ( self->ft_size ) {
        error = FT_Done_Size( self->ft_size );
        if(error) {
            freetype_error( error );
        }
    }
`

and then in texture_font_close, after the self->face = NULL:

`
self->ft_size = NULL;
`

As it is automatically disposed of by FT_Done_Face?

Not quite sure of the whole object graph here, so I haven't added it to my PR.

File: texture-font.c [hunk 3]
```diff
@@ -511,9 +542,16 @@ texture_font_delete( texture_font_t *self )
...
+    if(error) {
+	freetype_error( error, "FT_Error (%s:%d, code 0x%02x) : %s\n",
+			__FILENAME__, __LINE__, FT_Errors[error].code, FT_Errors[error].message);
+    }
```
Comment (michaeltyson) [inline texture-font.c]: Crash here in FT_Done_Size if texture_font_close is called before this point, which can happen if a load fails (happening to me right now as the emoji support isn't quite working yet).

Perhaps wrap it in a check:


```suggestion
    if ( self->ft_size ) {
        error = FT_Done_Size( self->ft_size );
        if(error) {
            freetype_error( error );
        }
    }
```

and then in texture_font_close, after the self->face = NULL:

```
self->ft_size = NULL;
```

As it is automatically disposed of by FT_Done_Face?

Not quite sure of the whole object graph here, so I haven't added it to my PR.

File: texture-font.c [hunk 4]
```diff
@@ -381,28 +393,54 @@ texture_font_load_face( texture_font_t *self, float size )
...
+	}
 
-    /* Set char size */
-    error = FT_Set_Char_Size(self->face, (int)(size * HRES), 0, DPI * HRES, DPI);
-    
-    if(error) {
-        freetype_error( error, "FT_Error (%s:%d, code 0x%02x) : %s\n",
-			__FILENAME__, __LINE__, FT_Errors[error].code, FT_Errors[error].message);
-        goto cleanup_face;
+	if( texture_is_color_font( self ) ) {
```
Comment (michaeltyson) [inline texture-font.c]: texture_is_color_font doesn't work for the Apple Emoji font. Here's what works for me:

```suggestion
      if( FT_HAS_FIXED_SIZES( self->face ) ) {
```

File: texture-font.c [hunk 5]
```diff
@@ -899,12 +892,30 @@ texture_font_load_glyph( texture_font_t * self,
...
-        dst_ptr += tgt_w * self->atlas->depth;
-        src_ptr += ft_bitmap.pitch;
+    if( self->atlas->depth == 4 ) {
+        for( i = 0; i < src_h; i++ ) {
+            int j;
+            // flip bgra to rgba, because that's better for OpenGL
+            for( j = 0; j < ft_bitmap.width; j++ ) {
+                uint32_t bgra, rgba;
+                bgra = ((uint32_t*)src_ptr)[j];
+#if __BYTE_ORDER == __BIG_ENDIAN
```
Comment (michaeltyson) [inline texture-font.c:914]: See my PR to your repo - __BYTE_ORDER and __BIG_ENDIAN are not defined macros in GCC, so this doesn't work on at least iOS.

File: texture-font.c [hunk 6]
```diff
@@ -249,6 +251,16 @@ texture_library_new()
...
+// -------------------------------------------------- texture_is_color_font ---
+
+int
+texture_is_color_font( texture_font_t *self) {
+    static const uint32_t tag = FT_MAKE_TAG('C', 'B', 'D', 'T');
+    unsigned long length = 0;
+    FT_Load_Sfnt_Table(self->face, tag, 0, NULL, &length);
+    return length != 0;
+}
+
```
[the remark this answers is not in the document]
Comment (michaeltyson) [reply texture-font.c]: Note: I have added this change and the one below to my PR to your repo

File: texture-font.c (hunk 4 above)
Comment (michaeltyson) [reply texture-font.c]: Note: I have added this change and the one above to my PR to your repo

**Comment by michaeltyson** [comment]
Lovely! These changes brought a reasonably completed scene's initial setup down from 0.32s to .18s.

File: texture-font.c (hunk 5 above)
Comment (forthy42) [reply texture-font.c:914]: __BYTE_ORDER is defined in <endian.h>, on Mac OS and iOS, this is <machine/endian.h>. I already check for

    #ifdef __APPLE__
    # include <machine/endian.h>

and this should do it.

Comment (forthy42) [reply texture-font.c:917]: Ok, looking at your suggested change, it's more complicated.  The people at Apple not only managed to put endian.h into a different location, they also managed to rename the macros.  Doh!  Your fix breaks Linux.

Solution is to first load Apple's endian.h and then redefine the Apple macros to match the Linux macros.

Comment (michaeltyson) [reply texture-font.c:917]: Ah! Those tricksy buggers!

**Comment by forthy42** [comment]
I tried Apple's icon fonts, and it works well when you check for the FT_HAS_FIXED_SIZES in both places where color fonts need special treatment.  Works with Android fonts (the one I tested this with originally), too.

**Comment by michaeltyson** [comment]
I'm still getting "unused function" build warnings on the rol8 function in texture-font.c (that's why I'd combined rol8/ror8 in my PR, rather than having the enddianness check at the point of calling). I could override the build warning, but it's probably better to just restructure a little, doncha reckon?

File: texture-font.c [hunk 7]
```diff
@@ -381,28 +393,54 @@ texture_font_load_face( texture_font_t *self, float size )
...
+		freetype_error( error, "FT_Error (%s:%d) : no fixed size in color font\n",
+				__FILENAME__, __LINE__);
+		goto cleanup_face;
+	    }
+	
+	    int best_match = 0;
+	    int diff = abs((int)size - self->face->available_sizes[0].width);
+
+	    for (int i = 1; i < self->face->num_fixed_sizes; ++i) {
+		int ndiff = abs((int)size - self->face->available_sizes[i].width);
```
Comment (michaeltyson) [inline texture-font.c]: This is resulting in emojis that are a bit too big: 
image

May I suggest only considering sizes that are less than or equal to the requested size? Also, is that i = 1 initial assignment deliberate? Seems odd.

```suggestion
	    for (i = 0; i < self->face->num_fixed_sizes; ++i) {
		if ( self->face->available_sizes[i].width > size ) continue;
		int ndiff = abs((int)size - self->face->available_sizes[i].width);
```

image

Comment (forthy42) [reply texture-font.c]: Yes, the `i=1` is intentional. I start with `best_match=0` (that's the index of the best match) and then try to find a better match.  There's no point to look at `i=0` again.  There's also no guarantee that there is a size available smaller than the one you requested.

The size chosen is the one with the smallest difference to the requested size, and you need to use the value of the new `self->scale` field to draw it correctly scaled.  It's set to 1.0 for normal fonts, so you can use this scaling factor all the time.

Comment (michaeltyson) [reply texture-font.c]: Ah! That makes sense!

Comment (forthy42) [reply texture-font.c]: The selection algorithm is from here.

https://github.com/hironishihara/ofxMixedFont/blob/master/src/ofxFT2Font.cpp

**Comment by forthy42** [comment]
> I'm still getting "unused function" build warnings on the rol8 function in texture-font.c

I changed that to a single rol(in, x) function; that should fix that problem.

**Comment by michaeltyson** [comment]
After that last rol() build warning, I think that's it from my end – everything's working great here. When that last thing's fixed, and when you're satisfied, @rougier, I think it's ready to merge 🎉

Thanks for all your great work, @forthy42!

image

**Comment by michaeltyson** [comment]
Yay! Okay, that’s it for me - whenever you’re ready, @rougier

**Comment by forthy42** [comment]
The only thing I've still a problem is `make test`.  First of all, the tga saved is upside down (which is to be expected from OpenGL), and needs a flip before comparing. So there needs to be some preprocessing, because `convert -flip` is much easier than affine transformation with `compare`. And then, a number of things have changed (upstream and in my code) that stops the images from comparing fine.

But I suggest to attack this part later.

**Comment by rougier** [comment]
Let's pull the trigger then! Many many thanks for your hard and nice work.

**Comment by Bromvlieg** [comment]
I've been using `FetchContent_Declare` with the repo url so I have latest version, but this merge seems to error the build :(
image

**Comment by rougier** [comment]
@Bromvlieg Thanks for the report
@forthy42 Do you know if this is related to your code?

**Comment by forthy42** [comment]
No idea, sorry.  This is a problem early on in the build process, and I don't even know why it expects a freetype-gld library instead of freetype-gl.  Typo?

**Comment by Bromvlieg** [comment]
The `d` at the end is for debug builds. When i try it with `release` it finds the target, but throws an error:

```
D:\code\cpp\mainframe\out\build\x64-Release\_deps\freetype-gl-src\ftgl-utils.h(22): fatal error C1075: '{': no matching token found
```

**Comment by forthy42** [comment]
Yes, this looks like the extern "C" is done twice. Can you delete the second one?  I think this all can go away:

```
#  ifdef __cplusplus
#   define FTGL_NEED_EXTERN_C
#ifndef IMPLEMENT_FREETYPE_GL
extern
#endif
"C" {
#  endif
```

**Comment by Bromvlieg** [comment]
I've removed the code and added it to the PR I've made.

**Pull request merged**